### PR TITLE
Return user friendly error message in error

### PR DESCRIPTION
**Pull request by b123400**
connect https://github.com/SkygearIO/skygear-SDK-Android/issues/89

This PR does not localise the error message, because I don't know how to do it in an Android library. Seems like I need to access the Context in order to get the resource file so I can access the localised string, but asking for Context in the error seems to be strange... Any one familiar with Android can help?

Update:
Localisation is going to be done in another issue

### CODE REVIEW

**Comment by b123400** [comment]
The push check failed on travis because I pushed to the wrong repo and then moved to my own, travis cannot handle that, it should be fine....

File: skygear/src/main/java/io/skygear/skygear/Error.java [hunk 1]
```diff
@@ -180,4 +180,48 @@ public static Code fromValue(int value) {
             return Code.UNEXPECTED_ERROR;
         }
     }
+
+    @Override
+    public String getLocalizedMessage() {
```
Comment (ben181231) [inline skygear/src/main/java/io/skygear/skygear/Error.java]: If we are not localized the error message, shall we just override the `getMessage()` ?

**Comment by cheungpat** [comment]
I will take over.

**Comment by b123400** [comment]
The issue is about providing both developer-facing and user-friendly error message. Since `getMessage()` has been used for developer-facing message and I don't want to introduce breaking changes, so I just used another method. Though I agree `getLocalizedMessage` is not an accurate name

**Comment by cheungpat** [comment]
I am going to propose the following interface changes for `io.skygear.skygear.Error`:

* `public int getCodeValue()` gets the integer value of the error code obtained from Server (`code` key). If the code value is not recognized by the class, the code value will save as is.
* `public int getName()` gets the string value of the error name obtained from Server (`name` key).
* `public String getDetailMessage()` gets the string message of the error obtained from the Server (`message` key). The error message is saved as is.
* `public Map<String, ?> getInfo()` gets the info dictionary of the error obtained from the Server (`info` key).
* Overrides `toString()` to print name, detail message and code.

Pseudo-implementation for constructor:

```
if codeValue is unspecified, codeValue = int(UNEXPECTED_ERROR)
if detailMessage is unspecified, detailMessage = ErrorMessage(codeValue)
if info is unspecified, info = Map() // empty Map
code = fromValue(codeValue) // if codeValue is unrecognized, fallback to UNEXPECTED_ERROR
message = ErrorMessage(code)
super(message)
save code, codeValue, detailMessage and info to instance variable
// (optional) save message to instance variable
```

Where:

* `ErrorMessage()` convert the error code to a predefined *English* error message

Therefore:

* `code.getValue() != codeValue` if `codeValue` is not recognized
* `getMessage()` is the *English* error message that can be shown to the user
* `getDetailMessage()` is the error message that *should not* be shown to the user

---

For localization:

I recognize that localization requires `android.content.Context`, and the context is best not referenced by `io.skygear.skygear.Error` because that requires importing a high-level class into a low-level class.

To work around this, we need a new class or method on an existing class that returns localized error message from the error code:

```
public String getLocalizedErrorMessage(io.skygear.skygear.Error error) { // not sure about the signature
  return Context.getString(R.string.<error name>)
}
```

I hope this makes sense.

**Comment by b123400** [comment]
Updated to provide `getDetailMessage` and `getInfo`. I don't think `getName` is needed because we have we already have code.

**Comment by cheungpat** [comment]
> I don't think getName is needed because we have we already have code.

... unless when the error code is new and the SDK is not updated to include the new error code > name mapping. The name in the error object will have the name of the new error.

**Comment by cheungpat** [comment]
You are missing this point:

> * code.getValue() != codeValue if codeValue is not recognized

The reason is the same as above... if the error code is new, and if the SDK is not updated, the app developer will not find out the real reason of the error.

**Comment by b123400** [comment]
OK, will add `getName`. The current implementation returns `UNEXPECTED_ERROR` when code is not recognised. It is done via `Code.fromValue`, the function returns `UNEXPECTED_ERROR` when the value is not recognised.

**Comment by cheungpat** [comment]
> It is done via Code.fromValue, the function returns UNEXPECTED_ERROR when the value is not recognised.

I think we need a new instance variable for storing the actual error code, do you agree?

**Comment by b123400** [comment]
OK, now it stores the original error code and error name.

**Pull request merged**